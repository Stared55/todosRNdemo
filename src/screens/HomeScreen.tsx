import React, { useEffect, useState } from 'react';
import {
  ActivityIndicator,
  FlatList,
  StyleSheet,
  Text,
  View,
} from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import Navbar from 'src/components/navigation/Navbar';
import ActionPanel from 'src/components/todos/ActionPanel';
import Todo from 'src/components/todos/Todo';
import { AppState } from 'src/store/store';
import { getTodos } from 'src/store/todos/todosActions';
import { Colors } from 'src/utils/styles/colors';
import { ITodo } from 'src/utils/types/todoTypes';
import { BottomTabNavigationProp } from '@react-navigation/bottom-tabs';
import { TabParamList } from 'src/components/navigation/RootNavigator';

type HomeScreenNavigationProp = BottomTabNavigationProp<TabParamList, 'Home'>;

export interface HomeScreenProps {
  navigation?: HomeScreenNavigationProp;
}

const HomeScreen: React.SFC<HomeScreenProps> = () => {
  const { todos, loading } = useSelector((state: AppState) => state.todos);
  const [items, setItems] = React.useState<ITodo[]>([]);
  const [selectedIds, setSelectedIds] = useState<number[]>([]);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getTodos());
  }, []);

  useEffect(() => {
    const itemsCopy = copyTodos();
    setItems(itemsCopy);
    renderList([...itemsCopy]);
  }, [todos]);

  const copyTodos = () => {
    if (todos?.length > 0) {
      return todos.map((todo: ITodo) => {
        return {
          ...todo,
          selected: false,
        };
      });
    } else {
      return [];
    }
  };

  const handleSelectedIds = (id: number): void => {
    let updatedSelectedIds = [...selectedIds];
    if (updatedSelectedIds.includes(id)) {
      updatedSelectedIds = updatedSelectedIds.filter(
        (selected) => selected !== id,
      );
    } else {
      updatedSelectedIds.push(id);
    }
    setSelectedIds(updatedSelectedIds);
  };

  const clearSelectedCards = () => {
    const itemsCopy = copyTodos();
    setItems([...itemsCopy]);
    renderList([...itemsCopy]);
    setSelectedIds([]);
  };

  const renderList = ({ item }: any) => {
    if (!item) {
      return <Text>No todos</Text>;
    }
    return (
      <Todo key={item.id} todo={item} setSelectedIds={handleSelectedIds} />
    );
  };

  const content =
    todos && todos.length > 0 ? (
      <FlatList
        style={styles.flatList}
        contentContainerStyle={styles.flatListContainer}
        data={items}
        keyExtractor={(item) => item.id.toString()}
        renderItem={renderList}
      />
    ) : (
      <View style={styles.noTodos}>
        <Text style={{}}>No todos found!</Text>
      </View>
    );

  return (
    <>
      <Navbar title="Home" />
      <ActionPanel clearSelected={clearSelectedCards} selected={selectedIds} />
      {loading ? (
        <ActivityIndicator />
      ) : (
        <View style={styles.wrapper}>{content}</View>
      )}
    </>
  );
};

export default HomeScreen;

const styles = StyleSheet.create({
  wrapper: {
    backgroundColor: Colors.white,
    width: '100%',
    justifyContent: 'flex-start',
    alignItems: 'center',
    flex: 1,
  },
  flatList: {
    width: '100%',
  },
  flatListContainer: {
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  noTodos: {
    alignItems: 'center',
    justifyContent: 'center',
  },
});
