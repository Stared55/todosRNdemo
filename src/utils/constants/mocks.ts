import { ITodo } from '../types/todoTypes';

export const mockedTodos: ITodo[] = [
  {
    id: 1,
    userId: 1,
    title: 'lorem 1',
    completed: true,
  },
  {
    id: 2,
    userId: 2,
    title: 'lorem 2',
    completed: false,
  },
  {
    id: 3,
    userId: 2,
    title: 'lorem 3',
    completed: true,
  },
];
