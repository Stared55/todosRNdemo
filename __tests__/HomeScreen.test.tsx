import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import HomeScreen from 'src/screens/HomeScreen';
import { mockedTodos } from 'src/utils/constants/mocks';

jest.mock('react-redux', () => ({
  useDispatch: () => () => jest.fn(),
  useSelector: (selector: any) =>
    selector({
      loading: false,
      todos: [],
      error: null,
    }),
}));

describe('HomeScreen', () => {
  let component: ShallowWrapper;
  let props: any;

  beforeEach(() => {
    component = shallow(<HomeScreen />);
    props = component.props();
    jest.clearAllMocks();
  });

  it('should match to snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('should render navbar with correct label', () => {
    const title = props.children[0].props.title;
    expect(title).toEqual('Home');
  });

  it('should render 3 children', () => {
    const children = props.children;
    expect(children).toHaveLength(3);
  });
});
