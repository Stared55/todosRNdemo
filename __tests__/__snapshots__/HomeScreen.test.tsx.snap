// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen should match to snapshot 1`] = `
<Fragment>
  <Navbar
    title="Home"
  />
  <ActionPanel
    clearSelected={[Function]}
    selected={Array []}
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "flex": 1,
        "justifyContent": "flex-start",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={Object {}}
      >
        No todos found!
      </Text>
    </View>
  </View>
</Fragment>
`;
